docs_src/design/adr/core/0019-EdgeX-CLI-V2.md

```diff
@@ -0,0 +1,113 @@
+| `-j`, `--json`   | output the raw JSON response returned by the EdgeX REST API and *nothing* else. This output mode is used for script-based usage of the client.    |
+| `--version`   | output the version of the client and if available, the version of EdgeX installed on the system (using the version of the core data service)   |
+
+
+2. Restructure the Go code hierarchy to follow the [most recent recommended guidelines](https://github.com/golang-standards/project-layout). For instance /cmd should just contain the main application for the project, not an implementation for each command - that should be in /internal/cmd
+
+3. Take full advantage of the features of the underlying command-line library, [Cobra](https://github.com/spf13/cobra), such as tab-completion of commands.
+
+4. Change the client binary name from `edgex-cli` to `edgex`. The reason is that the "-cli" is superfluous. There is no `edgex` tool and best practice is to choose a simple, memorable word. CLI tool names do not generally include `"-cli"` in the name. (Examples: `go`,`snap`,`kubectl`,`git`,`docker` etc)
+
+5. Add support for the four common endpoints - `/ping`, `/metrics`, `/version`, `/status`
+
+6. Change the syntax of the CLI to require a positional service arg which precedes the command, thus allowing overlap of command names across services. The service name can either be an abbreviation or the full name of the service . Examples:
+
+    - `edgex data ping`
+    - `edgex d ping`
+    - `edgex metadata version`
+    - `edgex m version`
+    - `edgex metadata device list`
+    - `edgex command device list`
+
+    An alternative solution is to use a positional argument only for the common endpoints but specify it after the command or to have a regular arg like `-s`/`--service` `<name>`).
+
+
+7. Implement all required V2 endpoints for core services
+
+    **Core Command** (`edgex (c)ommand`)
+    - **`command`** `get | put | list` *(we might want to replace get/put with read/write to be more consistent w/other commands)*
+
+    **Core Data** (`edgex (d)ata`)
+    - **`event`** `add | count | list | rm | scrub**`
+    - **`reading`** `count | list`
+
+    **Metadata** (`edgex (m)etadata`)
+    
+    - **`device`**  `add | adminstate | list | operstate | rm | update`
+    - **`deviceprofile`**  `add | list | rm | update`
+    - **`deviceservice`** ` add | list | rm | update`
+    - **`provisionwatcher`**  `add | list | rm | update`
+    
+    **Support Notifications** ( `edgex (n)otifications`)
+    - **`notification`** `add | list | rm`
+    - **`subscription`** `add | list | rm`
+
+   **Support Scheduler** ( `edgex (s)cheduler`)
+    - **`interval`** `add | list | rm | update` 
+
+
+    **Common endpoints in all services**
+    - **`version`**
+    - **`ping`**
+    - **`metrics`**
+    - **`status`** 
+
```

Comment:
I think there are some API endpoints/commands that are higher priority than others.  If we need to limit the functionality to some of the endpoints in this version, I am ok with that.  VMWare did a priority matrix that we can refer to if needed.

Reply:
@jpwhitemn - thank you for the comments. I have responded to them inline and updated the ADR.